Error Creating Athena Database

### Terraform Version
Terraform 0.11.3
aws 1.14.0

### Affected Resource(s)
Please list the resources as a list, for example:
- aws_athena_database

### Terraform Configuration Files
```hcl
resource "aws_athena_database" "pccp_datalake" {
  name          = "PccpDevDatalake"
  bucket        = "${aws_s3_bucket.pccp_datalake.bucket}"
  force_destroy = true
}
```

### Debug Output
https://gist.github.com/andrewmichael/4d1b1e5bf035b1e35264cd81111723c0

### Expected Behavior
Athena database is created, and subsequent plan / apply commands operate correctly.

### Actual Behavior
Plan output:-
```
Terraform will perform the following actions:

  + aws_athena_database.pccp_datalake
      id:            <computed>
      bucket:        "pccp-dev-datalake"
      force_destroy: "true"
      name:          "PccpDevDatalake"


Plan: 1 to add, 0 to change, 0 to destroy.
```
Apply end in error:-
```
Error: Error applying plan:

1 error(s) occurred:

* aws_athena_database.pccp_datalake: 1 error(s) occurred:

* aws_athena_database.pccp_datalake: [ERROR] Athena not found database: PccpDevDatalake, query result: default
pccpdevdatalake
```
Subsequent plan / apply fails.
```
Error: Error refreshing state: 1 error(s) occurred:

* aws_athena_database.pccp_datalake: 1 error(s) occurred:

* aws_athena_database.pccp_datalake: aws_athena_database.pccp_datalake: [ERROR] Athena not found database: PccpDevDatalake, query result: default
pccpdevdatalake
```
Observing the AWS console the database appears to have been created.
The only way out of this now is to manually deleting the Athena database, then revert to the previous version of remote state in S3.

Just reverting remote state, results in the following error:-
```
Error: Error applying plan:

1 error(s) occurred:

* aws_athena_database.pccp_datalake: 1 error(s) occurred:

* aws_athena_database.pccp_datalake: reason: FAILED: Execution Error, return code 1 from org.apache.hadoop.hive.ql.exec.DDLTask. Database PccpDevDatalake already exists

```

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. `terraform plan`
1. `terraform apply`

### Important Factoids
IAM user has admin permissions
remote state is persisted in s3

--- a/aws/resource_aws_athena_database.go
+++ b/aws/resource_aws_athena_database.go
@@ -23,11 +23,10 @@ func resourceAwsAthenaDatabase() *schema.Resource {
 
 		Schema: map[string]*schema.Schema{
 			"name": {
-				Type:     schema.TypeString,
-				Required: true,
-				ForceNew: true,
-				// based on https://docs.aws.amazon.com/athena/latest/ug/tables-databases-columns-names.html
-				ValidateFunc: validation.StringMatch(regexp.MustCompile("[_a-z0-9]+"), "see https://docs.aws.amazon.com/athena/latest/ug/tables-databases-columns-names.html"),
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringMatch(regexp.MustCompile("^[_a-z0-9]+$"), "see https://docs.aws.amazon.com/athena/latest/ug/tables-databases-columns-names.html"),
 			},
 			"bucket": {
 				Type:     schema.TypeString,

--- a/aws/resource_aws_athena_database.go
+++ b/aws/resource_aws_athena_database.go
@@ -2,8 +2,11 @@ package aws
 
 import (
 	"fmt"
+	"regexp"
 	"strings"
 	"time"
+
+	"github.com/hashicorp/terraform/helper/validation"
 
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/athena"
@@ -23,6 +26,8 @@ func resourceAwsAthenaDatabase() *schema.Resource {
 				Type:     schema.TypeString,
 				Required: true,
 				ForceNew: true,
+				// based on https://docs.aws.amazon.com/athena/latest/ug/tables-databases-columns-names.html
+				ValidateFunc: validation.StringMatch(regexp.MustCompile("[_a-z0-9]+"), "see https://docs.aws.amazon.com/athena/latest/ug/tables-databases-columns-names.html"),
 			},
 			"bucket": {
 				Type:     schema.TypeString,
@@ -42,7 +47,7 @@ func resourceAwsAthenaDatabaseCreate(d *schema.ResourceData, meta interface{}) e
 	conn := meta.(*AWSClient).athenaconn
 
 	input := &athena.StartQueryExecutionInput{
-		QueryString: aws.String(fmt.Sprintf("create database %s;", d.Get("name").(string))),
+		QueryString: aws.String(fmt.Sprintf("create database `%s`;", d.Get("name").(string))),
 		ResultConfiguration: &athena.ResultConfiguration{
 			OutputLocation: aws.String("s3://" + d.Get("bucket").(string)),
 		},
@@ -92,7 +97,7 @@ func resourceAwsAthenaDatabaseDelete(d *schema.ResourceData, meta interface{}) e
 	name := d.Get("name").(string)
 	bucket := d.Get("bucket").(string)
 
-	queryString := fmt.Sprintf("drop database %s", name)
+	queryString := fmt.Sprintf("drop database `%s`", name)
 	if d.Get("force_destroy").(bool) {
 		queryString += " cascade"
 	}

--- a/aws/resource_aws_athena_database_test.go
+++ b/aws/resource_aws_athena_database_test.go
@@ -30,6 +30,25 @@ func TestAccAWSAthenaDatabase_basic(t *testing.T) {
 	})
 }
 
+func TestAccAWSAthenaDatabase_nameStartsWithUnderscore(t *testing.T) {
+	rInt := acctest.RandInt()
+	dbName := acctest.RandString(8)
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSAthenaDatabaseDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAthenaDatabaseConfigNameStartsWithUnderscore(rInt, dbName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSAthenaDatabaseExists("aws_athena_database.hoge"),
+					resource.TestCheckResourceAttr("aws_athena_database.hoge", "name", "_"+dbName),
+				),
+			},
+		},
+	})
+}
+
 func TestAccAWSAthenaDatabase_destroyFailsIfTablesExist(t *testing.T) {
 	rInt := acctest.RandInt()
 	dbName := acctest.RandString(8)
@@ -248,7 +267,7 @@ func testAccCheckAWSAthenaDatabaseDropFails(dbName string) resource.TestCheckFun
 			QueryExecutionContext: &athena.QueryExecutionContext{
 				Database: aws.String(dbName),
 			},
-			QueryString: aws.String(fmt.Sprintf("drop database %s;", dbName)),
+			QueryString: aws.String(fmt.Sprintf("drop database `%s`;", dbName)),
 			ResultConfiguration: &athena.ResultConfiguration{
 				OutputLocation: aws.String("s3://" + bucketName),
 			},
@@ -292,6 +311,20 @@ func testAccAthenaDatabaseConfig(randInt int, dbName string) string {
 
     resource "aws_athena_database" "hoge" {
       name = "%s"
+      bucket = "${aws_s3_bucket.hoge.bucket}"
+    }
+    `, dbName, randInt, dbName)
+}
+
+func testAccAthenaDatabaseConfigNameStartsWithUnderscore(randInt int, dbName string) string {
+	return fmt.Sprintf(`
+    resource "aws_s3_bucket" "hoge" {
+      bucket = "tf-athena-db-%s-%d"
+      force_destroy = true
+    }
+
+    resource "aws_athena_database" "hoge" {
+      name = "_%s"
       bucket = "${aws_s3_bucket.hoge.bucket}"
     }
     `, dbName, randInt, dbName)

--- a/aws/resource_aws_athena_database_test.go
+++ b/aws/resource_aws_athena_database_test.go
@@ -2,6 +2,7 @@ package aws
 
 import (
 	"fmt"
+	"regexp"
 	"testing"
 
 	"github.com/aws/aws-sdk-go/aws"
@@ -21,7 +22,7 @@ func TestAccAWSAthenaDatabase_basic(t *testing.T) {
 		CheckDestroy: testAccCheckAWSAthenaDatabaseDestroy,
 		Steps: []resource.TestStep{
 			{
-				Config: testAccAthenaDatabaseConfig(rInt, dbName),
+				Config: testAccAthenaDatabaseConfig(rInt, dbName, false),
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckAWSAthenaDatabaseExists("aws_athena_database.hoge"),
 				),
@@ -32,18 +33,34 @@ func TestAccAWSAthenaDatabase_basic(t *testing.T) {
 
 func TestAccAWSAthenaDatabase_nameStartsWithUnderscore(t *testing.T) {
 	rInt := acctest.RandInt()
-	dbName := acctest.RandString(8)
+	dbName := "_" + acctest.RandString(8)
 	resource.Test(t, resource.TestCase{
 		PreCheck:     func() { testAccPreCheck(t) },
 		Providers:    testAccProviders,
 		CheckDestroy: testAccCheckAWSAthenaDatabaseDestroy,
 		Steps: []resource.TestStep{
 			{
-				Config: testAccAthenaDatabaseConfigNameStartsWithUnderscore(rInt, dbName),
+				Config: testAccAthenaDatabaseConfig(rInt, dbName, false),
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckAWSAthenaDatabaseExists("aws_athena_database.hoge"),
-					resource.TestCheckResourceAttr("aws_athena_database.hoge", "name", "_"+dbName),
+					resource.TestCheckResourceAttr("aws_athena_database.hoge", "name", dbName),
 				),
+			},
+		},
+	})
+}
+
+func TestAccAWSAthenaDatabase_nameCantHaveUppercase(t *testing.T) {
+	rInt := acctest.RandInt()
+	dbName := "A" + acctest.RandString(8)
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSAthenaDatabaseDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config:      testAccAthenaDatabaseConfig(rInt, dbName, false),
+				ExpectError: regexp.MustCompile(`see .*\.com`),
 			},
 		},
 	})
@@ -58,7 +75,7 @@ func TestAccAWSAthenaDatabase_destroyFailsIfTablesExist(t *testing.T) {
 		CheckDestroy: testAccCheckAWSAthenaDatabaseDestroy,
 		Steps: []resource.TestStep{
 			{
-				Config: testAccAthenaDatabaseConfig(rInt, dbName),
+				Config: testAccAthenaDatabaseConfig(rInt, dbName, false),
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckAWSAthenaDatabaseExists("aws_athena_database.hoge"),
 					testAccAWSAthenaDatabaseCreateTables(dbName),
@@ -79,7 +96,7 @@ func TestAccAWSAthenaDatabase_forceDestroyAlwaysSucceeds(t *testing.T) {
 		CheckDestroy: testAccCheckAWSAthenaDatabaseDestroy,
 		Steps: []resource.TestStep{
 			{
-				Config: testAccAthenaDatabaseConfigForceDestroy(rInt, dbName),
+				Config: testAccAthenaDatabaseConfig(rInt, dbName, true),
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckAWSAthenaDatabaseExists("aws_athena_database.hoge"),
 					testAccAWSAthenaDatabaseCreateTables(dbName),
@@ -100,7 +117,7 @@ func testAccCheckAWSAthenaDatabaseDestroy(s *terraform.State) error {
 		}
 
 		rInt := acctest.RandInt()
-		bucketName := fmt.Sprintf("tf-athena-db-%s-%d", rs.Primary.Attributes["name"], rInt)
+		bucketName := fmt.Sprintf("tf-athena-db-%d", rInt)
 		_, err := s3conn.CreateBucket(&s3.CreateBucketInput{
 			Bucket: aws.String(bucketName),
 		})
@@ -302,45 +319,17 @@ func testAccAthenaDatabaseFindBucketName(s *terraform.State, dbName string) (buc
 	return bucket, err
 }
 
-func testAccAthenaDatabaseConfig(randInt int, dbName string) string {
+func testAccAthenaDatabaseConfig(randInt int, dbName string, forceDestroy bool) string {
 	return fmt.Sprintf(`
     resource "aws_s3_bucket" "hoge" {
-      bucket = "tf-athena-db-%s-%d"
+      bucket = "tf-athena-db-%[1]d"
       force_destroy = true
     }
 
     resource "aws_athena_database" "hoge" {
-      name = "%s"
-      bucket = "${aws_s3_bucket.hoge.bucket}"
+      name = "%[2]s"
+	  bucket = "${aws_s3_bucket.hoge.bucket}"
+	  force_destroy = %[3]t
     }
-    `, dbName, randInt, dbName)
-}
-
-func testAccAthenaDatabaseConfigNameStartsWithUnderscore(randInt int, dbName string) string {
-	return fmt.Sprintf(`
-    resource "aws_s3_bucket" "hoge" {
-      bucket = "tf-athena-db-%s-%d"
-      force_destroy = true
-    }
-
-    resource "aws_athena_database" "hoge" {
-      name = "_%s"
-      bucket = "${aws_s3_bucket.hoge.bucket}"
-    }
-    `, dbName, randInt, dbName)
-}
-
-func testAccAthenaDatabaseConfigForceDestroy(randInt int, dbName string) string {
-	return fmt.Sprintf(`
-    resource "aws_s3_bucket" "hoge" {
-      bucket = "tf-athena-db-%s-%d"
-      force_destroy = true
-    }
-
-    resource "aws_athena_database" "hoge" {
-      name = "%s"
-      bucket = "${aws_s3_bucket.hoge.bucket}"
-	  force_destroy = true
-    }
-    `, dbName, randInt, dbName)
+    `, randInt, dbName, forceDestroy)
 }

